api/src/main/java/org/openmrs/module/locationbasedaccess/PatientSearchAdvise.java

```diff
@@ -0,0 +1,106 @@
+import org.springframework.aop.support.StaticMethodMatcherPointcutAdvisor;
+import java.lang.reflect.Method;
+import java.util.Iterator;
+import java.util.List;
+
+/**
+ * It will used to implement the AOP Methods for PatientServiceImpl class
+ */
+public class PatientSearchAdvise extends StaticMethodMatcherPointcutAdvisor implements Advisor {
+
+    protected static final Log log = LogFactory.getLog(PatientSearchAdvise.class);
+
+    @Override
+    public boolean matches(Method method, Class targetClass) {
+        // For PatientServiceImpl.savePatient()
+        if (method.getName().equals("getPatients")) {
+            return true;
+        }
+        // For PatientServiceImpl.getPatient()
+        else if (method.getName().equals("getPatient")) {
+            return true;
+        }
+        return false;
+    }
+
+    @Override
+    public Advice getAdvice() {
+        return new PatientSearchControlAdvice();
+    }
+
+    /**
+     * This inner class will be used to intercept the caught method using AOP
+     */
+    private class PatientSearchControlAdvice implements MethodInterceptor {
+
+        public Object invoke(MethodInvocation invocation) throws Throwable {
+            Object currentResult = invocation.proceed();
+            Integer sessionLocationId = Context.getUserContext().getLocationId();
+
+            if(sessionLocationId != null) {
+                Location sessionLocation = Context.getLocationService().getLocation(sessionLocationId);
+                // For PatientServiceImpl.getPatient()
+                if(currentResult instanceof Patient) {
+                    Patient currentPatient = (Patient)currentResult;
+                    currentPatient = checkPatientForAccessLocation(currentPatient, sessionLocation);
+                    return  currentPatient;
+                }
+                // For PatientServiceImpl.savePatient()
+                else {
+                    List<Patient> currentPatientResults = (List<Patient>)currentResult;
+                    for (Iterator<Patient> iterator = currentPatientResults.iterator(); iterator.hasNext(); ) {
+                        Patient patient  = iterator.next();
+                        patient = checkPatientForAccessLocation(patient, sessionLocation);
+                        if(patient == null) {
+                            iterator.remove();
+                        }
+                    }
+                    return currentPatientResults;
+                }
+            }
+            else {
+                // If session location doesn't available, return empty patient
+                log.debug("Search Patient : Null Session Location in the UserContext");
+                return  null;
+            }
+        }
+
+        private  Patient checkPatientForAccessLocation(Patient patient, Location sessionLocation) {
+            Person person = patient.getPerson();
+            PersonAttribute personAttribute = person.getAttribute("accessControlLocation");
```

Comment:
Can you use a constant for the above to make the code more readable?

Reply:
Sure, I will add